### PR TITLE
[YUNIKORN-1820] update protoc, protoc-gen-go, grpc

### PR DESCRIPTION
**Pull request by FrankYang0529**
### What is this PR for?

Version change:
* protoc: 23.2
* protoc-gen-go: latest release version
* protoc-gen-go-grpc: 1.3.0
* grpc: 1.56.0

I replace yunikorn-scheduler-interface version in both yunikorn-core & yunikorn-k8shim with `YUNIKORN-1820` branch. In yunikorn-k8shim, I can run `make build` and `make test` without error. I will send another two PRs to update core & k8shim after this PR is merged.
* https://github.com/FrankYang0529/yunikorn-k8shim/tree/YUNIKORN-1820
* https://github.com/FrankYang0529/yunikorn-core/tree/YUNIKORN-1820

### What type of PR is it?
* [ ] - Bug Fix
* [X] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?

https://issues.apache.org/jira/browse/YUNIKORN-1820

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -22,23 +22,21 @@ module github.com/apache/yunikorn-scheduler-interface
...
+	google.golang.org/grpc v1.56.0
+	google.golang.org/protobuf v1.30.0
 )
 
 require (
-	golang.org/x/sys v0.6.0 // indirect
-	golang.org/x/text v0.8.0 // indirect
-	google.golang.org/genproto v0.0.0-20180817151627-c66870c02cf8 // indirect
+	golang.org/x/net v0.9.0 // indirect
+	golang.org/x/sys v0.7.0 // indirect
```
Comment (wilfred-s) [inline go.mod:32]: The latest ones I can see are:
```
golang.org/x/net => golang.org/x/net v0.10.0
golang.org/x/sys => golang.org/x/sys v0.8.0
golang.org/x/tools => golang.org/x/tools v0.9.3
```

File: go.mod [hunk 2]
```diff
@@ -22,23 +22,21 @@ module github.com/apache/yunikorn-scheduler-interface
...
+	golang.org/x/text v0.9.0 // indirect
+	google.golang.org/genproto v0.0.0-20230410155749-daa745c078e1 // indirect
 )
 
 replace (
```
Comment (wilfred-s) [inline go.mod:37]: Can we also update these replace values?
Keep them insync with the ones above etc

**Review by wilfred-s — CHANGES_REQUESTED** [review]
minor import changes, base looks good

**Comment by FrankYang0529** [comment]
Hi @wilfred-s, I updated the PR to include `protoc-gen-go-grpc` to fix the CI. I also update related core & k8shim to check it can work.
* core: https://github.com/FrankYang0529/yunikorn-core/commit/083fbff4794e1c7134249ee12c872d058627c8ba
* k8shim: https://github.com/FrankYang0529/yunikorn-k8shim/commit/30f85239d7d7e267a14cf1272909a553a450890a

File: lib/go/Makefile [hunk 3]
```diff
@@ -70,29 +77,46 @@ $(PROTOC):
...
+	stat $@ > /dev/null 2>&1
+
+# Get and install the protoc-gen-go-grpc for protoc.
+PROTOC_GEN_GO_GRPC_TAR := protoc-gen-go-grpc.v$(PROTOC_GEN_GO_GRPC_VER).$(PROTOGEN_OS).$(PROTOGEN_ARCH).tar.gz
+PROTOC_GEN_GO_GRPC_URL := https://github.com/grpc/grpc-go/releases/download/cmd/protoc-gen-go-grpc/v$(PROTOC_GEN_GO_GRPC_VER)/$(PROTOC_GEN_GO_GRPC_TAR)
+PROTOC_GEN_GO_GRPC     := $(PROTOC_BIN_DIR)/protoc-gen-go-grpc
+$(PROTOC_GEN_GO_GRPC):
+	mkdir -p $(PROTOC_TMP_DIR) && \
+		curl -L $(PROTOC_GEN_GO_GRPC_URL) -o $(PROTOC_TMP_DIR)/$(PROTOC_GEN_GO_GRPC_TAR) && \
+		tar -C $(PROTOC_BIN_DIR) -xzvf $(PROTOC_TMP_DIR)/$(PROTOC_GEN_GO_GRPC_TAR) protoc-gen-go-grpc && \
```
Comment (wilfred-s) [inline lib/go/Makefile:98]: This fails on linux as the file has a path specified in the tar file:
```
tar tfv protoc/protoc-gen-go-grpc.v1.3.0.linux.amd64.tar.gz
drwxr-xr-x  0 runner docker      0  2 Mar 10:51 ./
-rw-r--r--  0 runner docker   4863  2 Mar 10:51 ./README.md
-rwxr-xr-x  0 runner docker 8344207  2 Mar 10:51 ./protoc-gen-go-grpc
-rw-r--r--  0 runner docker   11358  2 Mar 10:51 ./LICENSE
```
That needs to be reflected in the tar command. Linux is not as forgiving as MacOSX is when unpacking

**Review by wilfred-s — CHANGES_REQUESTED** [review]

**Comment by wilfred-s** [comment]
The update of the scheduler interface needs to include some of the changes that @pbacsko has made for the event system. We need to do that in one go as you have noticed the build will otherwise break. We can track the changes for those other repositories as one update as part of this jira YUNIKORN-1820.

File: lib/go/Makefile (hunk 3 above)
Comment (FrankYang0529) [reply lib/go/Makefile]: Thanks. Updated it!

File: go.mod [hunk 4]
```diff
@@ -22,24 +22,23 @@ module github.com/apache/yunikorn-scheduler-interface
...
+	github.com/golang/protobuf v1.5.3 // indirect
+	golang.org/x/net v0.10.0 // indirect
+	golang.org/x/sys v0.8.0 // indirect
+	golang.org/x/text v0.9.0 // indirect
+	google.golang.org/genproto v0.0.0-20230410155749-daa745c078e1 // indirect
 )
 
 replace (
-	github.com/golang/protobuf => github.com/golang/protobuf v1.2.0
 	golang.org/x/crypto => golang.org/x/crypto v0.7.0
```
[the remark this answers is not in the document]
Comment (FrankYang0529) [reply go.mod]: Added it. Thanks.

**Comment by pbacsko** [comment]
Tried it under Ubuntu, it works.

```
~/repos/yunikorn-scheduler-interface$ make
make -C /home/bacskop/repos/yunikorn-scheduler-interface/ build
make[1]: Entering directory '/home/bacskop/repos/yunikorn-scheduler-interface'
cat scheduler-interface-spec.md | sed -n -e '/```protobuf$/,/^```$/ p' | sed '/^```/d' >> si.proto.tmp
awk '{ if (length > 200) print NR, $0 }' si.proto.tmp | diff - /dev/null
(diff si.proto.tmp si.proto > /dev/null 2>&1 || mv -f si.proto.tmp si.proto) && \
	rm -f si.proto.tmp
test -d lib/go/common || mkdir -p lib/go/common
cat scheduler-interface-spec.md | sed -n -e '/```constants$/,/^```$/ p' | sed '/^```/ s/.*//g' >> lib/go/common/tmp_constants.go
go fmt lib/go/common/tmp_constants.go
lib/go/common/tmp_constants.go
(diff lib/go/common/tmp_constants.go lib/go/common/constants.go > /dev/null 2>&1 || mv -f lib/go/common/tmp_constants.go lib/go/common/constants.go) && \
	rm -f lib/go/common/tmp_constants.go
test -d lib/go/api || mkdir -p lib/go/api
cat scheduler-interface-spec.md | sed -n -e '/```golang$/,/^```$/ p' | sed '/^```/ s/.*//g' >> lib/go/api/tmp_interface.go
go fmt lib/go/api/tmp_interface.go
go: downloading google.golang.org/protobuf v1.30.0
go: downloading google.golang.org/grpc v1.56.0
go: downloading google.golang.org/genproto v0.0.0-20230410155749-daa745c078e1
go: downloading golang.org/x/net v0.10.0
go: downloading golang.org/x/sys v0.8.0
go: downloading golang.org/x/text v0.9.0
lib/go/api/tmp_interface.go
(diff lib/go/api/tmp_interface.go lib/go/api/interface.go > /dev/null 2>&1 || mv -f lib/go/api/tmp_interface.go lib/go/api/interface.go) && \
	rm -f lib/go/api/tmp_interface.go
make -C lib/go
make[2]: Entering directory '/home/bacskop/repos/yunikorn-scheduler-interface/lib/go'
mkdir -p protoc && \
	curl -L https://github.com/grpc/grpc-go/releases/download/cmd/protoc-gen-go-grpc/v1.3.0/protoc-gen-go-grpc.v1.3.0.linux.amd64.tar.gz -o protoc/protoc-gen-go-grpc.v1.3.0.linux.amd64.tar.gz && \
	tar -C protoc/bin -xzvf protoc/protoc-gen-go-grpc.v1.3.0.linux.amd64.tar.gz ./protoc-gen-go-grpc && \
	chmod 0755 protoc/bin/protoc-gen-go-grpc
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 4328k  100 4328k    0     0  5770k      0 --:--:-- --:--:-- --:--:-- 25.1M
./protoc-gen-go-grpc
stat protoc/bin/protoc-gen-go-grpc > /dev/null 2>&1
ls -al protoc/bin
total 19100
drwxr-xr-x 2 bacskop bacskop    4096 jún   20 07:59 .
drwxrwxr-x 4 bacskop bacskop    4096 jún   20 07:59 ..
-rwxr-xr-x 1 bacskop bacskop 5463176 ápr   21  2022 protoc
-rwxrwxr-x 1 bacskop bacskop 5735469 jún   14 11:27 protoc-gen-go
-rwxr-xr-x 1 bacskop bacskop 8344207 márc   2 00:51 protoc-gen-go-grpc
mkdir -p si/build && \
	(cd /home/bacskop/repos && /home/bacskop/repos/yunikorn-scheduler-interface/lib/go/protoc/bin/protoc \
	--go_out=/home/bacskop/repos/yunikorn-scheduler-interface/lib/go/si/build --go_opt=paths=source_relative \
	--go-grpc_out=/home/bacskop/repos/yunikorn-scheduler-interface/lib/go/si/build --go-grpc_opt=paths=source_relative \
	yunikorn-scheduler-interface/si.proto)

syntax check for constants code

protobuf go source generated:
	/home/bacskop/repos/yunikorn-scheduler-interface/lib/go/si/si.pb.go
	/home/bacskop/repos/yunikorn-scheduler-interface/lib/go/si/si_grpc.pb.go

make[2]: Leaving directory '/home/bacskop/repos/yunikorn-scheduler-interface/lib/go'
make[1]: Leaving directory '/home/bacskop/repos/yunikorn-scheduler-interface'
```

**Review by pbacsko — APPROVED** [review]
+1 LGTM

**Review by wilfred-s — APPROVED** [review]
LGTM

**Comment by pbacsko** [comment]
@FrankYang0529 please create the PRs for the shim & core. Thanks.

**Pull request closed without merging**